### PR TITLE
NEW Create DBClassNameVarchar

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/11358

### CODE REVIEW

File: src/ORM/DatabaseAdmin.php [hunk 1]
```diff
@@ -460,7 +461,8 @@ protected function getClassNameRemappingFields()
         foreach ($dataClasses as $className) {
             $fieldSpecs = $schema->fieldSpecs($className);
             foreach ($fieldSpecs as $fieldName => $fieldSpec) {
-                if (Injector::inst()->create($fieldSpec, 'Dummy') instanceof DBClassName) {
+                $dummy = Injector::inst()->create($fieldSpec, 'Dummy');
+                if ($dummy instanceof DBClassName || $dummy instanceof DBClassNameVarchar) {
```
Comment (emteknetnz) [inline src/ORM/DatabaseAdmin.php:465]: This is the only logic I found that specifically care if a field is a DBClassName ... and this code is only normally called by DatabaseAdmin::migrateClassNames() which uses the config `classname_value_remapping` - which is only used when migrating sites from Silverstripe 3, so it's largely irrelevant.

I've created a new card to remove https://github.com/silverstripe/silverstripe-framework/issues/11362 this legacy functionality

Comment (GuySartorelli) [reply src/ORM/DatabaseAdmin.php:465]: > which is only used when migrating sites from Silverstripe 3

It's useful for migrating data at any time in a project's lifetime - we should make sure this functionality is still working.

File: src/ORM/FieldType/DBClassNameVarchar.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * This will use more disk space, though will prevent issues with long dev/builds on
+ * very large database tables when a ALTER TABLE queries are required to update the enum.
+ *
+ * Use the following config to use this class in your project:
+ *
+ * ```
+ * SilverStripe\ORM\DataObject:
+ *   fixed_fields:
+ *     ClassName: DBClassNameVarchar
+ * ```
```
Comment (GuySartorelli) [inline src/ORM/FieldType/DBClassNameVarchar.php]:
```suggestion
 * <code>
 * SilverStripe\ORM\DataObject:
 *   fixed_fields:
 *     ClassName: DBClassNameVarchar
 * </code>
```
Needs to be HTML markup for api.silverstripe.org to read it correctly.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
@emteknetnz I saw you doing some performance tests of enum vs varchar, but I don't see the results of that anyway. Can you please dump those in a comment somewhere so I can validate your results?

File: src/ORM/FieldType/DBClassNameVarchar.php (hunk 2 above)
Comment (emteknetnz) [reply src/ORM/FieldType/DBClassNameVarchar.php]: Done

**Comment by emteknetnz** [comment]
I used the code in https://github.com/silverstripe/silverstripe-framework/issues/11358#issuecomment-2325394074 slightly modified to generate 1,000,000 rows and to set a fairly standard looking FQCN

```
// ...
        $num = 1000000; // total number of records that should exist (note cannot decrease)
        $insert = 100000; // max inserts per query
        $count = self::get()->count();
        $diff = $num - $count;
        $t = '2024-08-29 17:47:10';
        parent::requireDefaultRecords();
        if ($diff > 0) {
            # use raw SQL to make insertions much faster
            $loops = ceil($diff / $insert);
            for ($loop = 0; $loop < $loops; $loop++) {
                $sqlBase = [];
                $sqlA = [];
                for ($i = 1; $i <= $insert && $i <= $diff; $i++) {
                    $id = $count + ($loop * $insert) + $i;
                    $sqlBase[] = "($id, 'App\\\\Src\\\\Something\\\\MyDataObjectA', 'My Data Object $id', '$t', '$t')";
                    $sqlA[] = "($id, 'Some value $id')";
                }
                if (!empty($sqlBase)) {
                    DB::query('INSERT INTO "MyBaseDataObject" ("ID", "ClassName", "Title", "LastEdited", "Created") VALUES ' . implode(',', $sqlBase) . ';');
                    DB::query('INSERT INTO "MyDataObjectA" ("ID", "SomeField") VALUES ' . implode(',', $sqlA) . ';
                    ');
                }
                $diff -= $insert;
            }
        }
    }
// ...
```

I then used the config to switch between enum and varchar, and then ran the following SQL to get the database size - note you may need to drop the database before switching as the ClassName isn't a valid class

```
SELECT table_schema "DB Name", ROUND(SUM(data_length + index_length) / 1024 / 1024, 1) "DB Size in MB" FROM information_schema.tables GROUP BY table_schema;

enum
+--------------------+---------------+
| DB Name            | DB Size in MB |
+--------------------+---------------+
| information_schema |           0.2 |
| mysql              |          10.3 |
| performance_schema |           0.0 |
| SS_mysite          |         118.6 |
+--------------------+---------------+

varchar
+--------------------+---------------+
| DB Name            | DB Size in MB |
+--------------------+---------------+
| information_schema |           0.2 |
| mysql              |          10.3 |
| performance_schema |           0.0 |
| SS_mysite          |         180.8 |
+--------------------+---------------+
4 rows in set (0.01 sec)
```

At first glance it's a huge relative % increase, though ignore that because there's almost no other data in the database so the ClassName column is a large proportion of it. In the real world this won't be the case

Size difference is 180.8 - 118.6 = 62.2mb per 1,000,000, or 6.2 mb per 100,000.

I realised the first time I did that I didn't have quad slashes in my script so the slashes weren't in the database so I had slightly less data stored. I'll bump up the estimate to 7mb per 100,000 in the docs to be conservative

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
After I've changed to varchar, if I add or remove a `DataObject` subclass, I still see the following in `dev/build` which are clearly alter table queries:

```text
Field ChangeSetItem.ObjectClass: changed to enum( ........ ) character set utf8mb4 collate utf8mb4_unicode_ci default 'SilverStripe\\Assets\\File' (from enum( ....... )

Field FileLink.ParentClass: changed to enum( ........ ) character set utf8mb4 collate utf8mb4_unicode_ci default 'SilverStripe\\Assets\\File' (from enum( ....... )

Field SiteTreeLink.ParentClass: changed to enum( ........ ) character set utf8mb4 collate utf8mb4_unicode_ci default 'SilverStripe\\Assets\\File' (from enum( ....... )
```

The first one is particularly concerning, because change set items are created every time anything is published, which means there will be a _lot_ of those records.
We should probably either account for that somehow in the code, or else add examples of how to update these as well in the docs.

**Comment by emteknetnz** [comment]
Need some additional config to get DBPolymorphicForeignKey to use varchar, have updated code sample + docs

**Review by GuySartorelli — APPROVED** [review]
LGTM

**Pull request merged**